### DISCUSSION TITLE
Porting macOS aliases from `find` to `fd`

### DISCUSSION THREAD

**Discussion by shamindras** [sharkdp/fd#1615] [Q&A] [answered]
I'm really enjoying using `fd`!

I'm in the process of porting over a couple of useful `macOS` clean up aliases from `find` to `fd`. These aliases are as follows:

```bash
alias clean="find . -maxdepth 1 -name '*.DS_Store' -o -name '*~' -o -name '.*~' -o -name '*.swo' -o -name '*.swp' -o -name '.*.swo' | xargs rm"

alias wash="find . -name '*.DS_Store' -o -name '*~' -o -name '.*~' -o -name '*.swo' -o -name '*.swp' -o -name '.*.swo' | xargs rm"
```

When switching `find` to `fd` these aliases don't work. I tried several variations, but they don't even filter the multiple file type matches correctly. I also noticed that once ported to `fd`, we can replace the `xargs` with [command execution](https://github.com/sharkdp/fd?tab=readme-ov-file#command-execution), i.e., `-x` in `fd`. But wasn't sure how to port it over.

I would also like to ensure that the revised `fd` versions of these commands only filter for **files**, not directories.

Could anyone please assist with porting these commands?

**Comment by tavianator** [comment]
`find -maxdepth 1` is equivalent to `fd --max-depth 1`.  `-name X -o -name Y` can be written like `fd 'X|Y'`, but they have to be regexes.  So something like this:

```console
$ fd '\.(DS_Store|swo|swp)$|~$' -d1 -X rm
```

By the way your `find` commands have a bit of redundancy: `-name '*~'` will already match anything that `-name '.*~'` does, and same with `*.swo` and `.*.swo`.  You could also use `-exec` instead of `xargs` but it is more awkward with `find` syntax:

```console
$ find . -maxdepth 1 \( -name '*.DS_Store' -o -name ... \) -exec rm {} +
```

**Reply by shamindras** [reply]
Thanks @tavianator for your helpful guidance. I tried testing your command as follows:

```bash
mkdir -p test_fd
cd test_fd

# Create nested directories
mkdir -p sampdir{1..5}/subdir{1..3}

# Create test files for each level of the nested directory tree
touch {a.swo,b.swo,.DS_Store,c.swp,d.swp,aa~,bb~}
touch sampdir{1..5}/{a.swo,b.swo,.DS_Store,c.swp,d.swp,aa~,bb~}
touch sampdir{1..5}/subdir{1..3}/{a.swo,b.swo,.DS_Store,c.swp,d.swp,aa~,bb~}
tree --level=3
```

Which produces this directory test case

<details><summary>Toggle me!</summary>

```bash
.
├── sampdir1
│  ├── subdir1
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir2
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir3
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── .DS_Store
│  ├── a.swo
│  ├── aa~
│  ├── b.swo
│  ├── bb~
│  ├── c.swp
│  └── d.swp
├── sampdir2
│  ├── subdir1
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir2
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir3
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── .DS_Store
│  ├── a.swo
│  ├── aa~
│  ├── b.swo
│  ├── bb~
│  ├── c.swp
│  └── d.swp
├── sampdir3
│  ├── subdir1
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir2
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir3
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── .DS_Store
│  ├── a.swo
│  ├── aa~
│  ├── b.swo
│  ├── bb~
│  ├── c.swp
│  └── d.swp
├── sampdir4
│  ├── subdir1
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir2
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir3
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── .DS_Store
│  ├── a.swo
│  ├── aa~
│  ├── b.swo
│  ├── bb~
│  ├── c.swp
│  └── d.swp
├── sampdir5
│  ├── subdir1
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir2
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── subdir3
│  │  ├── .DS_Store
│  │  ├── a.swo
│  │  ├── aa~
│  │  ├── b.swo
│  │  ├── bb~
│  │  ├── c.swp
│  │  └── d.swp
│  ├── .DS_Store
│  ├── a.swo
│  ├── aa~
│  ├── b.swo
│  ├── bb~
│  ├── c.swp
│  └── d.swp
├── .DS_Store
├── a.swo
├── aa~
├── b.swo
├── bb~
├── c.swp
└── d.swp
```

</details>


Yet when I run `fd '\.(DS_Store|swo|swp)$|~$' -d1` or `fd '\.(DS_Store|swo|swp)$|~$'`, I don't seem to get any matches. Did I make a mistake here?

Also how does your regex ensure, for example, that we match `*.swo` (desired) and not just `.swo` (without the wildcard prefix, i.e., undesired)?

**Reply by tavianator** [reply, marked as answer]
> Yet when I run `fd '\.(DS_Store|swo|swp)$|~$' -d1` or `fd '\.(DS_Store|swo|swp)$|~$'`, I don't seem to get any matches. Did I make a mistake here?

You probably want to add `fd -u` to include hidden files and `.gitignore`d files.  Also note that the `-d1`/`--max-depth=1` is in effect, remove that for wash.

> Also how does your regex ensure, for example, that we match `*.swo` (desired) and not just `.swo` (without the wildcard prefix, i.e., undesired)?

`fd`'s regexes are not anchored, so `fd X` is just like `fd '.*X'`.

**Reply by shamindras** [reply]
Thanks @tavianator - your suggestions worked!

> You probably want to add fd -u to include hidden files and .gitignored files. Also note that the -d1/--max-depth=1 is in effect, remove that for wash.

This is great! Since the test directory in my previous was not a git repo, I was wondering why the hidden files were being ignored. I realize that `fd` is looking up my _global_ gitignore here. So I add the `-u` flag per your suggestion, which made it work.

> fd's regexes are not anchored, so `fd X` is just like `fd '.*X'`.

Thanks - this makes sense now.

For ease of reference here are the final ported commands you have provided for the convenience of future readers. Note that I added the `-tf` flag, to ensure that the removal applies _only_ to files (per my original requirement):

```bash
# clean up current directory
alias clean="fd '\.(DS_Store|swo|swp)$|~$' -tf -d1 -u -X rm"

# clean up current directory AND all subdirectories
alias wash="fd '\.(DS_Store|swo|swp)$|~$' -tf -u -X rm"
```
